<div align="center">
    <h1>
    CenProteo: Finding the Essential Proteins in a Protein Interaction Network
    </h1>
    <p>
    Project of BIO2502 Programming Languages for Bioinformatics, 2024 Spring, SJTU
    <br />
    <a href="https://github.com/xywawawa"><strong>xywawawa</strong></a>
    &nbsp;
    <a href="https://github.com/Cannizzaro-reaction"><strong>Cannizzaro-reaction</strong></a>
    &nbsp;
    <a href="https://github.com/Imiloin"><strong>Imiloin</strong></a>
    &nbsp;
    </p>
    <p>
    <a href="https://github.com/Imiloin/CenProteo"><img alt="Github Repository" src="https://img.shields.io/badge/Github-Repository-blue?logo=github&logoColor=blue"></a>
    <a href="https://github.com/Imiloin/CenProteo?tab=MIT-1-ov-file"><img alt="mit" src="https://img.shields.io/badge/License-MIT-red.svg"></a>
    </p>
    <img src="README.assets/asb_network_blue.svg" alt="asb_network_blue" width="200"/>
</div>

åœ¨è¿‡å»çš„å‡ åå¹´ä¸­ï¼Œå¯¹äºå•ä¸€è›‹ç™½è´¨çš„æ€§è´¨åŠåŠŸèƒ½æ–¹é¢çš„ç ”ç©¶å–å¾—äº†å¾ˆå¤§è¿›å±•ã€‚ä½†æ˜¯ï¼Œè›‹ç™½è´¨åœ¨ç”Ÿç‰©ä½“å†…å¾ˆå°‘å•ç‹¬å‘æŒ¥ä½œç”¨ï¼Œå› æ­¤äº†è§£è›‹ç™½è´¨ä¹‹é—´çš„ç›¸äº’ä½œç”¨å¯¹äºæ­ç¤ºå¤æ‚åˆ†å­æœºåˆ¶è‡³å…³é‡è¦ã€‚è¿‘å¹´æ¥ï¼Œé…µæ¯åŒæ‚äº¤ç³»ç»Ÿï¼ˆYeast Two-Hybrid, Y2Hï¼‰ï¼Œäº¤å‰é“¾æ¥è´¨è°±æ³•ï¼ˆCross-linking Mass Spectrometry, XL-MSï¼‰ç­‰é«˜é€šé‡å®éªŒæŠ€æœ¯å¿«é€Ÿå‘å±•ï¼Œä½¿å¾—è¶Šæ¥è¶Šå¤šè›‹ç™½è´¨ä¹‹é—´çš„ç›¸äº’ä½œç”¨è¢«ç ”ç©¶å’Œå‘è¡¨ï¼Œä¹Ÿç§¯ç´¯äº†å¤§é‡çš„ç›¸å…³å®éªŒæ•°æ®ï¼Œç”±æ­¤æ„å»ºå‡ºè›‹ç™½è´¨ç›¸äº’ä½œç”¨ç½‘ç»œï¼ˆProteinâ€“Protein Interactions Network, PPINï¼‰ ã€‚åœ¨ PPIN ä¸­ï¼Œå…³é”®è›‹ç™½å…·æœ‰ç‰¹å®šçš„æ‹“æ‰‘ä½ç½®å’ŒåŠŸèƒ½è§’è‰²ï¼Œå¯¹ç»´æŒç½‘ç»œçš„ç¨³å®šæ€§å’ŒåŠŸèƒ½å…·æœ‰é‡è¦å½±å“ã€‚ä¸ºäº†ä» PPIN ä¸­å‘ç°å…³é”®è›‹ç™½ï¼Œä¸€ç³»åˆ—å¦‚åº¦ä¸­å¿ƒæ€§ï¼ˆDegree Centralityï¼‰ï¼Œä»‹æ•°ä¸­å¿ƒæ€§ï¼ˆBetweenness Centralityï¼‰ï¼Œèšç±»ç³»æ•°ï¼ˆClustering Coefficientï¼‰ç­‰å›¾è®ºä¸­çš„ä¼ ç»Ÿç®—æ³•è¢«åº”ç”¨åˆ° PPIN ä¸­ï¼Œä½†è¿™äº›ç®—æ³•å¾€å¾€å¿½ç•¥äº†è›‹ç™½è´¨çš„åŠŸèƒ½ä½œç”¨ï¼Œä»…å…³æ³¨ç½‘ç»œæ‹“æ‰‘ç»“æ„ã€‚

è¿‘å¹´çš„ä¸€äº›æ–‡çŒ®æå‡ºäº†ä¸€äº›æ–°çš„ç®—æ³•ï¼Œè¯•å›¾ä½¿ç”¨æ›´å¤šçš„ç”Ÿç‰©æ•°æ®æ¥æé«˜å…³é”®è›‹ç™½çš„é¢„æµ‹å‡†ç¡®æ€§ã€‚è¿™äº›ç®—æ³•åŒ…æ‹¬äº†åŸºå› è¡¨è¾¾é‡æ•°æ®ã€äºšç»†èƒå®šä½æ•°æ®ã€åŸºå› åŒæºæ€§æ•°æ®ç­‰ã€‚ä½¿ç”¨æ›´å¤šçš„ç”Ÿç‰©å­¦æ•°æ®ï¼Œå¯ä»¥æ›´å¥½åœ°åæ˜ è›‹ç™½è´¨åœ¨ç”Ÿç‰©ä½“å†…çš„åŠŸèƒ½ä½œç”¨ï¼Œä»è€Œæé«˜å…³é”®è›‹ç™½çš„é¢„æµ‹å‡†ç¡®æ€§ã€‚

æœ¬é¡¹ç›®æ„å»ºäº†åŒ… `cenproteo` ï¼Œå®ç°äº†è‹¥å¹²è®¡ç®—è›‹ç™½è´¨ç½‘ç»œä¸­å„è›‹ç™½è´¨çš„ä¸­å¿ƒæ€§å¹¶è¿›è¡Œæ’åºä»è€Œå¯»æ‰¾å…³é”®è›‹ç™½è´¨çš„ç®—æ³•ï¼ŒéªŒè¯ç®—æ³•çš„å‡†ç¡®æ€§å¹¶è¿›è¡Œäº†æ•ˆæœæ¯”å¯¹ã€‚



## ğŸ—‚ï¸ Data Source & Preprocessing

åœ¨ `cenproteo` åŒ…å®ç°çš„å‡ ç§ç®—æ³•ä¸­ï¼Œä½¿ç”¨äº†é…¿é…’é…µæ¯ï¼ˆSaccharomyces cerevisiaeï¼‰çš„è›‹ç™½è´¨äº’ä½œä¿¡æ¯ï¼Œä¸»è¦ç”¨åˆ°äº†ä»¥ä¸‹å‡ ç§æ•°æ®ï¼š

* è›‹ç™½è´¨äº’ä½œå¯¹åŠè›‹ç™½è´¨ GO è¯­ä¹‰ç›¸ä¼¼æ€§æ•°å€¼ï¼ˆThe GO similarity value for BP, MF, and CC under the DIP PPI dataset and the combined PPI datasetï¼‰ï¼š

    åŸºå› æœ¬ä½“è®ºæœ¯è¯­ï¼ˆGene Ontology termï¼‰æ˜¯ç”Ÿç‰©ä¿¡æ¯å­¦ä¸­ç”¨æ¥æ ‡å‡†åŒ–åŸºå› äº§å“å±æ€§çš„ä¸€ç§æ–¹å¼ï¼Œå…è®¸ç ”ç©¶äººå‘˜ä»¥ä¸€ç§æ ‡å‡†åŒ–çš„æ–¹å¼æ³¨é‡ŠåŸºå› å’Œè›‹ç™½è´¨çš„åŠŸèƒ½ï¼Œæœ‰åŠ©äºæ•°æ®çš„å…±äº«å’Œæ¯”è¾ƒã€‚GO æœ¯è¯­è¦†ç›–äº†ä¸‰ä¸ªä¸»è¦é¢†åŸŸï¼šåˆ†å­åŠŸèƒ½ï¼ˆMolecular Functionï¼ŒMFï¼‰ï¼Œç»†èƒç»„åˆ†ï¼ˆCellular Componentï¼ŒtCCï¼‰å’Œç”Ÿç‰©è¿‡ç¨‹ï¼ˆBiological Processï¼ŒBPï¼‰ã€‚

    æ•°æ®æ¥è‡ªæ–‡çŒ® [zhang2016topology][1] æ”¯æ’‘ææ–™ã€‚

* åŸºå› è¡¨è¾¾é‡æ•°æ®ï¼š

    åŸºå› è¡¨è¾¾é‡æ•°æ®æ˜¯æŒ‡åœ¨ç‰¹å®šæ—¶é—´ã€ç‰¹å®šæ¡ä»¶æˆ–ç‰¹å®šç»†èƒç±»å‹ä¸­ï¼Œå„ä¸ªåŸºå› äº§ç”Ÿçš„ RNA åˆ†å­æ•°é‡çš„æµ‹é‡æ•°æ®ã€‚ç”±äºå…·æœ‰ä¸€å®šç›¸ä¼¼æ€§çš„è›‹ç™½æ›´æœ‰å¯èƒ½å…±è¡¨è¾¾ï¼Œå¯ä»¥åˆ©ç”¨åŸºå› è¡¨è¾¾é‡æ•°æ®è®¡ç®—çš®å°”é€Šç›¸å…³ç³»æ•°ï¼ˆPCCï¼‰ï¼Œä»è€Œè¡¨æ˜ä¸€å¯¹è›‹ç™½çš„ç›¸å…³æ€§å¼ºå¼±ã€‚

    æ•°æ®é€šè¿‡æ–‡çŒ® [benjamin2005metabolic][2] æä¾›çš„æ•°æ®ç¼–å·ï¼ˆGSE3431ï¼‰ï¼Œä» NCBI ä¸­ä¸‹è½½ç›¸åº”çš„åŸºå› è¡¨è¾¾é‡æ•°æ®ã€‚å¤„ç†æ–¹æ³•å¦‚ä¸‹ï¼š

    * ä¸‹è½½åŸºå› è¡¨è¾¾é‡åŸæ–‡ä»¶ï¼Œåˆ å»æ•°æ®ä»¥å¤–çš„å¤šä½™éƒ¨åˆ†ï¼Œå¹¶å°†æ•°æ®å†™å…¥ `.csv` æ–‡ä»¶ï¼›
    * å¯¹æ•°æ®è¿›è¡Œè¿‡æ»¤ï¼Œå¦‚æœè¯¥è›‹ç™½ä¸å­˜åœ¨äºé…¿é…’é…µæ¯äº’ä½œè›‹ç™½å¯¹ä¸­ï¼Œåˆ™åˆ å»ï¼Œå…¶ä½™ä¿ç•™ï¼›
    * é€šè¿‡ GSE3431 æ•°æ®é¡µé¢è¿›å…¥æµ‹åºå¹³å° GPL90 ä¿¡æ¯é¡µé¢ï¼Œä¸‹è½½è¯¥åŸºå› è¡¨è¾¾é‡æµ‹å®šçš„æ³¨é‡Šæ–‡ä»¶ï¼Œé€šè¿‡è„šæœ¬å°†åŸè¿‡æ»¤æ–‡ä»¶ä¸­çš„é€šé“ç¼–å·æ›¿æ¢ä¸ºå¯¹åº”çš„è›‹ç™½è´¨ç¼–å·ï¼Œå¦‚ä¸å­˜åœ¨åˆ™åˆ å»ã€‚

* äºšç»†èƒå®šä½æ•°æ®ï¼š

    ç ”ç©¶è¡¨æ˜ï¼Œå¾ˆå¤šäº’ä½œè›‹ç™½å¯¹éƒ½å­˜åœ¨ä¸ç»†èƒä¸­çš„åŒä¸€åˆ†åŒºæˆ–é‚»è¿‘åˆ†åŒºå†…ã€‚å› æ­¤äºšç»†èƒå®šä½æ•°æ®å¯¹äºå‘ç°å…³é”®è›‹ç™½æœ‰ä¸€å®šå¸®åŠ©ã€‚

    æ•°æ®æ¥è‡ª [COMPARTMENT database](https://compartments.jensenlab.org/Downloads)ï¼Œé€‰æ‹© `All channels integrated` ä¸­çš„ `yeast` é€‰é¡¹è¿›è¡Œä¸‹è½½ã€‚å¤„ç†æ–¹æ³•å¦‚ä¸‹ï¼š

    * å°†åŸå§‹æ•°æ®ç²˜è´´å…¥ `.csv` æ–‡æ¡£ï¼›
    * é€šè¿‡ 11 ä¸ªäºšç»†èƒå®šä½åˆ†åŒºæ‰€å¯¹åº”çš„ GO æœ¯è¯­ï¼Œå¯¹æ•°æ®è¿›è¡Œç­›é€‰ï¼Œå°†ç¬¦åˆè¿™ 11 ä¸ª GO æœ¯è¯­çš„æ•°æ®ä¿å­˜åˆ°æ–°çš„ `.csv` æ–‡ä»¶ï¼Œç”¨äºåç»­è®¡ç®—ã€‚

* åŸºå› åŒæºæ€§æ•°æ®ï¼š

    ç ”ç©¶å‘ç°ï¼Œè›‹ç™½è´¨çš„å…³é”®ç¨‹åº¦ä¸å…¶åœ¨è¿›åŒ–ä¸­çš„ä¿å®ˆæ€§æœ‰ä¸€å®šå…³ç³»ã€‚å¦‚æœä¸€ä¸ªè›‹ç™½åœ¨å¤šä¸ªç‰©ç§ä¸­å…·æœ‰åŒæºè›‹ç™½ï¼Œåˆ™è¯¥è›‹ç™½æ›´æœ‰å¯èƒ½ä¸ºå…³é”®è›‹ç™½ã€‚

    æ•°æ®æ¥è‡ª [InParanoid database (version 7)](https://inparanoid8.sbc.su.se/download/old_versions/data_7.0/)ï¼Œé€‰æ‹© `sqltables.tgz` è¿›è¡Œä¸‹è½½ï¼Œåç»­å¤„ç†å¦‚ä¸‹ï¼š

    * è§£å‹æ–‡ä»¶ï¼Œåˆ©ç”¨è„šæœ¬é€‰æ‹©ä¸é…¿é…’é…µæ¯æœ‰å…³çš„æ•°æ®ä¿å­˜ï¼Œå†™å…¥ `.csv` æ–‡ä»¶ï¼Œåˆ å»å…¶ä½™æ–‡ä»¶ï¼›
    * é€šè¿‡è„šæœ¬éå†ç­›é€‰å¾—åˆ°çš„æ¯ä¸ªæ–‡ä»¶ï¼Œç­›é€‰åŒæºæ€§ 100% çš„åŸºå› ç¼–å·ä¿å­˜ï¼Œå¹¶è®¡ç®—æ¯ä¸€ç§åŒæºåŸºå› ç¼–å·åœ¨ä¸åŒç‰©ç§ä¸­å‡ºç°çš„æ¬¡æ•°ï¼Œè®¡å…¥ `.csv` æ–‡æ¡£ã€‚

* å·²çŸ¥å…³é”®è›‹ç™½è¡¨ï¼šæ•°æ®æ¥è‡ª [DEG database](https://tubic.org/deg/public/index.php/query/eukaryotes/degac/DEG2001.html?lineage=eukaryotes&field=degac&term=DEG2001&page=1)ï¼Œé€‰æ‹© `Download > Eukaryotes > Organisms`ï¼Œä¸‹è½½ `Saccharomyces cerevisiae` çš„å…³é”®è›‹ç™½ä¿¡æ¯ã€‚æ•°æ®å¤„ç†è¿‡ç¨‹å¦‚ä¸‹ï¼š

    * é€šè¿‡ python è„šæœ¬ï¼Œä»åŸæ–‡ä»¶ä¸­æå–å…³é”®è›‹ç™½åç§°ï¼Œå†™å…¥æ–°çš„ `.csv` æ–‡ä»¶ï¼›

    * ä» [STRING database](https://cn.string-db.org/cgi/input?sessionId=bWGl1KdZES6m&input_page_show_search=on) ä¸­è·å–ä¸¤ç§è›‹ç™½è´¨ï¼ˆåŸºå› ï¼‰ç¼–ç æ–¹å¼çš„åç§°å¯¹ç…§è¡¨ï¼Œå¹¶åˆ©ç”¨è„šæœ¬å¾—åˆ°å¯¹åº”çš„è›‹ç™½åç§°ï¼Œå†™å…¥ `.csv` æ–‡æ¡£ï¼›
    * åˆ©ç”¨å¾—åˆ°çš„å¯¹åº”åç§°ï¼Œå¯¹å…³é”®è›‹ç™½çš„åç§°è¿›è¡Œæ›´æ–°ï¼ˆåŒæ—¶åŒ…å«ä¸¤ç§å‘½åæ¨¡å¼ï¼‰ï¼Œä¾¿äºåç»­åœ¨ç®—æ³•ä¸­ä½¿ç”¨ã€‚



## ğŸ—ï¸ Algorithms

è›‹ç™½è´¨ç½‘ç»œé€šå¸¸è¡¨ç¤ºä¸ºä¸€ä¸ªæ— å‘å›¾ $G=(V, E)$ï¼ŒèŠ‚ç‚¹ $u\in V$ è¡¨ç¤ºä¸€ä¸ªè›‹ç™½è´¨ï¼Œè¾¹ $(u,v) \in E$ è¡¨ç¤ºä¸¤ä¸ªè›‹ç™½è´¨ä¹‹é—´çš„ç›¸äº’ä½œç”¨ã€‚æˆ‘ä»¬ç”¨ $N$ è¡¨ç¤ºå›¾ä¸­èŠ‚ç‚¹æ€»æ•°ï¼Œ $A$ è¡¨ç¤ºå›¾çš„é‚»æ¥çŸ©é˜µã€‚

æ ¹æ®ä½¿ç”¨çš„æ•°æ®ç±»å‹ï¼Œç®—æ³•å¯ä»¥å¤§è‡´åˆ†ä¸ºä»¥ä¸‹å‡ ç±»ï¼š

#### ä¼ ç»Ÿç®—æ³•

ä»…ä½¿ç”¨ç½‘ç»œæ‹“æ‰‘æ•°æ®ï¼ˆ`cenproteo` ä¸­å®ç°çš„ `classical algortihms`ï¼‰è®¡ç®—è›‹ç™½è´¨çš„ä¸­å¿ƒæ€§ï¼š

+ DCï¼ˆdegree centralityï¼‰åº¦ä¸­å¿ƒæ€§ï¼šä¸€ä¸ªèŠ‚ç‚¹ $u$ çš„åº¦ä¸­å¿ƒæ€§ $DC(u)$ æ˜¯å…¶è¿æ¥çš„è¾¹æ•°ã€‚

    $$DC(u) = \sum_{v} a_{u,v}$$

+ BCï¼ˆBetweenness Centralityï¼‰ä»‹æ•°ä¸­å¿ƒæ€§ï¼šä¸€ä¸ªèŠ‚ç‚¹ $u$ çš„ä»‹æ•°ä¸­å¿ƒæ€§ $BC(u)$ å®šä¹‰ä¸ºé€šè¿‡èŠ‚ç‚¹ $u$ çš„æœ€çŸ­è·¯å¾„çš„å¹³å‡æ¯”ä¾‹ã€‚

    $$BC(u) = \sum_{s} \sum_{t} \frac{\rho(s, u, t)}{\rho(s, t)}, \quad s \neq t \neq u$$

    $\rho(s, t)$ æŒ‡çš„æ˜¯ $s$ å’Œ $t$ ä¹‹é—´çš„æœ€çŸ­è·¯å¾„æ•°ç›®ï¼Œ $\rho(s, u, t)$ æŒ‡çš„æ˜¯ $s$ å’Œ $t$ ä¹‹é—´çš„æœ€çŸ­è·¯å¾„ä¸­ç»è¿‡ $u$ çš„æ•°ç›®ã€‚

+ ECï¼ˆEigenvector Centralityï¼‰ç‰¹å¾å‘é‡ä¸­å¿ƒæ€§ï¼šä¸€ä¸ªèŠ‚ç‚¹ $u$ çš„ç‰¹å¾å‘é‡ä¸­å¿ƒæ€§ $EC(u)$ å®šä¹‰ä¸º $A$ çš„ä¸»ç‰¹å¾å‘é‡çš„ç¬¬ $u$ åˆ†é‡ã€‚

    $$EC(u) = \alpha_{\max}(u)$$

    $\alpha_{\max}$ æŒ‡çš„æ˜¯ $A$ çš„æœ€å¤§å€¼å¯¹åº”çš„ç‰¹å¾å‘é‡ï¼Œ $\alpha_{\max}(u)$ æŒ‡çš„æ˜¯ $\alpha_{\max}$ çš„ç¬¬ $u$ ä¸ªåˆ†é‡ã€‚

+ SCï¼ˆSubgraph Centralityï¼‰å­å›¾ä¸­å¿ƒæ€§ï¼šä¸€ä¸ªèŠ‚ç‚¹ $u$ çš„å­å›¾ä¸­å¿ƒæ€§ $SC(u)$ è¡¡é‡çš„æ˜¯èŠ‚ç‚¹ $u$ å‚ä¸çš„æ•´ä¸ªç½‘ç»œä¸­å­å›¾çš„æ•°é‡ã€‚

    $$SC(u) = \sum_{l=0}^{\infty} \frac{\mu_{l}(u)}{l!}$$

    $\mu_{l}(u)$ æŒ‡çš„æ˜¯å¼€å§‹å¹¶ç»“æŸäºèŠ‚ç‚¹ $u$ ä¸”é•¿åº¦ä¸º $l$ çš„ç¯è·¯æ•°ç›®ã€‚

+ ICï¼ˆInformation Centralityï¼‰ä¿¡æ¯ä¸­å¿ƒæ€§ï¼šä¸€ä¸ªèŠ‚ç‚¹ $u$ çš„ä¿¡æ¯ä¸­å¿ƒæ€§ $IC(u)$ è¡¡é‡çš„æ˜¯ä»¥èŠ‚ç‚¹ $u$ ç»“æŸçš„è·¯å¾„é•¿åº¦çš„è°ƒå’Œå¹³å‡å€¼ã€‚

    $$IC(u) = \left[\frac{1}{N} \sum_{v} \frac{1}{I_{uv}}\right]^{-1},I_{uv} = (c_{uu} + c_{vv} - c_{uv})^{-1},C = (c_{uv}) = [D - A + J]^{-1}$$

    $D$ ä¸ºæ¯ä¸ªèŠ‚ç‚¹åº¦çš„å¯¹è§’çŸ©é˜µï¼Œ $C$ æ˜¯æ”¹è¿›çš„é‚»æ¥çŸ©é˜µï¼Œ $J$ æ˜¯æ‰€æœ‰å…ƒç´ éƒ½ä¸º 1 çš„çŸ©é˜µã€‚

    åœ¨ `cenproteo` ä¸­ï¼Œä¸ºç®€åŒ–è®¡ç®—ï¼Œä¿¡æ¯ä¸­å¿ƒæ€§é€šè¿‡è®¡ç®— `curent flow centrality` æ¥è¿‘ä¼¼ã€‚

+ CCï¼ˆCloseness Centralityï¼‰æ¥è¿‘ä¸­å¿ƒæ€§ï¼šä¸€ä¸ªèŠ‚ç‚¹ $u$ çš„æ¥è¿‘ä¸­å¿ƒæ€§ $CC(u)$ æ˜¯ä»èŠ‚ç‚¹ $u$ åˆ°ç½‘ç»œä¸­æ‰€æœ‰å…¶ä»–èŠ‚ç‚¹çš„å›¾ç†è®ºè·ç¦»ä¹‹å’Œçš„å€’æ•°ã€‚

    $$CC(u) = \frac{N - 1}{\sum_{v} d(u, v)}$$

    $d(u,v)$ æŒ‡çš„æ˜¯ä»èŠ‚ç‚¹ $u$ åˆ°ç»“ç‚¹ $n$ çš„è·ç¦»ã€‚

+ NCï¼ˆNeighbor Centralityï¼‰é‚»å±…ä¸­å¿ƒæ€§ï¼šèŠ‚ç‚¹ $u$ çš„é‚»åŸŸä¸­å¿ƒæ€§ $NC(u)$ å®šä¹‰ä¸ºèŠ‚ç‚¹ $u$ é‚»å±…ä¹‹é—´çš„è¾¹èšç±»ç³»æ•°ï¼ˆEdge Clustering Coefficient, ECCï¼‰ä¹‹å’Œã€‚

    $$NC(u) = \sum_{v \in N_u} ECC(u, v),$$ 
    $$ECC(u, v) = \frac{z_{u, v}}{\min(d_u - 1, d_v - 1)},$$
    $$z_{u, v} = \sum_{w} A_{uw} A_{vw}.$$

    è¾¹èšç±»ç³»æ•° $ECC(u, v)$ è¡¨ç¤ºèŠ‚ç‚¹ $u$ å’ŒèŠ‚ç‚¹ $v$ ä¹‹é—´çš„å…±åŒé‚»å±…æ•° $z_{u, v}$ ä¸ä¸¤è€…åº¦æ•°çš„æœ€å°å€¼ä¹‹æ¯”ï¼Œ $A_{uw}$ å’Œ $A_{vw}$ åˆ†åˆ«è¡¨ç¤ºèŠ‚ç‚¹ $u$ å’Œ $v$ æ˜¯å¦ä¸èŠ‚ç‚¹ $w$ ç›¸è¿ã€‚

   

#### ç°ä»£ç®—æ³•

ä½¿ç”¨ç½‘ç»œæ‹“æ‰‘æ•°æ®å’ŒåŸºå› è¡¨è¾¾é‡ç­‰ç”Ÿç‰©æ•°æ®ï¼Œä¸»è¦å®ç°ä»¥ä¸‹å‡ ç§æ–¹æ³•ï¼š

+ TGSO algorithmï¼š

    <img src="README.assets/tgso.jpeg"/>

    è®¡ç®—æµç¨‹æ¡†æ¶ï¼š

    ```mermaid
    flowchart TD
        A[Initialize S, lambda, p] --> B[Compute ADN]
        A --> C[Compute PCC]
        A --> D[Compute CEN]
        A --> E[Compute CLN]
        B --> F[Combine results to compute initial matrix]
        C --> F
        D --> F
        E --> F
        F --> G[Compute LSG]
        G --> H[Initialize IS]
        H --> I[Iterate to find k seeds]
        I --> J[Update S]
    ```

    reference: [li2021iteration][3]ã€‚

+ JDC algorithmï¼š

    <img src="README.assets/jdc.jpeg"/>

    reference: [zhong2021novel][4]ã€‚

+ TEO algorithmï¼š

    <img src="README.assets/teo.jpeg" />

    reference: [zhang2016topology][1]ã€‚



## ğŸ”§ Installation

æœ¬ä»“åº“æä¾›äº† `cenprotro` çš„æºä»£ç ä¾›å®‰è£…ã€‚

#### Clone this repo

```bash
git clone https://github.com/Imiloin/CenProteo.git
cd CenProteo
```

#### Setup

```bash
pip install -e .
```



## â™¾ï¸ Usage

#### classical algorithms
* å¯¼å…¥ classical_algorithms æ¨¡å—ï¼Œé€‰æ‹©ç®—æ³•ï¼ˆè¿™é‡Œä»¥ DC ä¸ºä¾‹ï¼‰å¹¶è®¡ç®—å¾—åˆ°æ’åºåç½‘ç»œä¸­æ‰€æœ‰è›‹ç™½è´¨çš„ä¸­å¿ƒæ€§åˆ†æ•°ï¼š
    ```python
    from cenproteo import classical_algorithms
    class_test = classical_algorithms(<path_to_ppi_file>)
    dc_sorted_score = class_test.DC()
    ```

    * `path/to/ppi_file.csv`: PPI æ–‡ä»¶çš„è·¯å¾„ï¼ŒCSV æ–‡ä»¶åº”è‡³å°‘åŒ…å«å‰ä¸¤åˆ—ï¼Œè¡¨ç¤ºè›‹ç™½è´¨é—´çš„ç›¸äº’ä½œç”¨ã€‚ä¾‹å¦‚ï¼š
        ```
        Protein A,Protein B
        P1,P2
        P2,P3
        ```
        
* å°†ç»“æœå­˜å‚¨ä¸º `.csv` æ–‡ä»¶ï¼š
    ```python
    class_test.export_result_to_csv(dc_sorted_score, <path_to_save_result>)
    ```
  
* å¦‚æœæœ‰é‡‘æ ‡å‡†æ–‡ä»¶ï¼Œå°†ç®—æ³•å¾—åˆ°çš„å…³é”®è›‹ç™½è´¨ä¸é‡‘æ ‡å‡†è¿›è¡Œæ¯”è¾ƒï¼Œè¾“å‡ºnä¸ªå…³é”®è›‹ç™½è´¨ä¸­é¢„æµ‹æ­£ç¡®çš„ä¸ªæ•°ï¼š
    ```python
    class_test.first_n_comparison(n, dc_sorted_score, <path_to_real_essential_protein_file>)
    ```

    * `path_to_real_essential_protein_file.csv`: é‡‘æ ‡å‡†å…³é”®è›‹ç™½è´¨æ–‡ä»¶çš„è·¯å¾„ï¼ŒCSV æ–‡ä»¶åº”è‡³å°‘åŒ…å«ä¸€åˆ—ï¼Œåˆ—å‡ºå…³é”®è›‹ç™½è´¨çš„åç§°ï¼Œä¾‹å¦‚ï¼š
        ```
        CommonName,SystematicName
        AAR2,YBL074C
        ABD1,YBR236C
        ABF1,YKL112W
        ```

#### JDC algorithm, TGSO algorithm and TEO algorithm
ä»¥ JDC æ¨¡å—ä¸ºä¾‹ï¼ŒTGSO æ¨¡å—å’Œ TEO æ¨¡å—ç”¨æ³•ä¸ JDC æ¨¡å—ç›¸åŒã€‚
* å¯¼å…¥ JDC æ¨¡å—ï¼Œè¿ç”¨ JDC ç®—æ³•è®¡ç®—å¾—åˆ°æ’åºåç½‘ç»œä¸­æ‰€æœ‰è›‹ç™½è´¨çš„ JDC ä¸­å¿ƒæ€§åˆ†æ•°ï¼š
    ```python
    from cenproteo import JDC
    jdc_test =  JDC(<path_to_ppi_file>, <path_to_gene_expression_file>)
    jdc_sorted_score = jdc_test.calculate_jdc()
    ```
    
    * `path/to/ppi_file.csv`: PPI æ–‡ä»¶çš„è·¯å¾„ï¼ŒCSV æ–‡ä»¶åº”è‡³å°‘åŒ…å«å‰ä¸¤åˆ—ï¼Œè¡¨ç¤ºè›‹ç™½è´¨é—´çš„ç›¸äº’ä½œç”¨ã€‚ä¾‹å¦‚ï¼š
        ```
        protein A,protein B
        P1,P2
        P2,P3
        ```
    
    * `path/to/gene_expression_file.csv`: åŸºå› è¡¨è¾¾æ–‡ä»¶çš„è·¯å¾„ï¼ŒCSV æ–‡ä»¶åº”åŒ…å«å¤šåˆ—ï¼Œç¬¬ä¸€åˆ—ä¸ºåŸºå› IDï¼Œåç»­åˆ—ä¸ºåŸºå› åœ¨ä¸åŒæ ·æœ¬ä¸­çš„è¡¨è¾¾æ°´å¹³ï¼Œæœ€åä¸¤åˆ—ä¸ºå‡å€¼å’Œæ–¹å·®ã€‚ä¾‹å¦‚ï¼š
        ```
        ID_REF,GSM77298,GSM77299,...,mean,std
        YLR331C,0.055618618,0.073988438,...,0.071587456,0.033760167
        YLR332W,1.994324565,1.805780292,...,1.81596206,0.816943903
        ```
  
* å°†ç»“æœå­˜å‚¨ä¸º `.csv` æ–‡ä»¶ï¼š
    ```python
    jdc_test.export_result_to_csv(<path_to_save_result>)
    ```

* å¦‚æœæœ‰é‡‘æ ‡å‡†æ–‡ä»¶ï¼Œå°†ç®—æ³•å¾—åˆ°çš„å…³é”®è›‹ç™½è´¨ä¸é‡‘æ ‡å‡†è¿›è¡Œæ¯”è¾ƒï¼Œè¾“å‡º n ä¸ªå…³é”®è›‹ç™½è´¨ä¸­é¢„æµ‹æ­£ç¡®çš„ä¸ªæ•°ï¼š
    ```python
    jdc_test.first_n_comparsion(n, <path_to_real_essential_protein_file>)
    ```


## ğŸ“ˆ Results & Comparison

å°† `cenproteo` ä¸­çš„å‡ ç§ç®—æ³•è¿›è¡Œå¯¹æ¯”ï¼Œé€‰æ‹©ä¸åŒçš„ n è¿›è¡Œæµ‹è¯•æ—¶ï¼Œå¾—åˆ†æœ€é«˜çš„å‰ n ä¸ªè›‹ç™½ä¸­æ­£ç¡®çš„å…³é”®è›‹ç™½æ•°é‡å¦‚ä¸‹ï¼š

<img src="README.assets/comparison.svg" alt="img"  />

åˆ†åˆ«å– N = 100ï¼Œ200ï¼Œ400 è¿›è¡Œå¯¹æ¯”ï¼š

<img src="README.assets/different_n_counts.svg" alt="img" style="zoom:80%;" />

å…¶ä¸­ TGSO ç®—æ³•çš„æ­£ç¡®æ€§ç•¥é«˜äºå…¶ä»–ç®—æ³•ï¼Œä¸”å½“æ‰€é€‰å–çš„ n å€¼è¾ƒå°æ—¶ï¼Œå„ç®—æ³•çš„å‡†ç¡®æ€§ç›¸å¯¹è¾ƒé«˜ã€‚



## ğŸ“„ References

\[1\]  Zhang W, Xu J, Li Y, *et al*. Detecting essential proteins based on network topology, gene expression data, and gene ontology information. *IEEE/ACM transactions on computational biology and bioinformatics*, 2016, 15(1): 109-116. DOI: [10.1186/s12859-021-04175-8](https://doi.org/10.1186/s12859-021-04175-8)

\[2\]  Benjamin P. Tu et al. ,Logic of the Yeast Metabolic Cycle: Temporal Compartmentalization of Cellular Processes.Science310,1152-1158(2005). DOI: [10.1126/science.1120499](https://doi.org/10.1126/science.1120499)

\[3\]  Li S, Zhang Z, Li X, *et al*. An iteration model for identifying essential proteins by combining comprehensive PPI network with biological information. *BMC bioinformatics*, 2021, 22: 1-25. DOI: [10.1186/s12859-021-04300-7](https://doi.org/10.1186/s12859-021-04300-7)

\[4\]  Zhong, J., Tang, C., Peng, W. *et al.* A novel essential protein identification method based on PPI networks and gene expression data. *BMC Bioinformatics* 22, 248 (2021). DOI: [10.1186/s12859-021-04175-8](https://doi.org/10.1186/s12859-021-04175-8)



[1]: https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04175-8 "zhang2016topology"

[2]: https://doi.org/10.1126/science.1120499 "benjamin2005metabolic"

[3]: https://doi.org/10.1186/s12859-021-04300-7 "li2021iteration"

[4]: https://doi.org/10.1186/s12859-021-04175-8 "zhong2021novel"
